server/channels/app/channel_bookmark.go

```diff
@@ -130,3 +134,51 @@ func (a *App) UpdateChannelBookmarkSortOrder(bookmarkId, channelId string, newIn
 
 	return bookmarks, nil
 }
+
+func (a *App) GetChannelsWithBookmarksForSession(c request.CTX, session *model.Session, channels model.ChannelList, since int64) []*model.ChannelWithBookmarks {
+	channelIds := []string{}
+	for _, channel := range channels {
+		if a.SessionHasPermissionToChannel(c, *session, channel.Id, model.PermissionReadChannelContent) {
+			channelIds = append(channelIds, channel.Id)
+		}
+	}
+
+	if bookmarksMap, err := a.GetAllChannelBookmarks(channelIds, since); err == nil {
+		channelsWithBookmarks := []*model.ChannelWithBookmarks{}
+		for _, channel := range channels {
+			cb := &model.ChannelWithBookmarks{
+				Channel:   channel,
+				Bookmarks: bookmarksMap[channel.Id],
+			}
+			channelsWithBookmarks = append(channelsWithBookmarks, cb)
+		}
+
+		return channelsWithBookmarks
```

Comment:
This means that you could pass a list of 10 channels to this method and return only 8 based on the user permissions, right? If the user initially had permissions to retrieve those channels, I think we should preserve them in the response, even if they didn't get the bookmarks included

Reply:
not really, we are iterating over the **channels** that we passed as an argument and then adding to the slice that gets returned. if the channel in the slice does not have bookmarks, then the property gets set as `nil`